File: 2. Add Two Numbers/addTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *i=l1,*j=l2;
        ListNode *answer=0,*list;
        list = new ListNode(0);
        int add=0,a,b,sum;
        while (i || j) {
            if (i) {
                a=i->val;
                i=i->next;
            }
            else    a=0;
            if (j) {
                b=j->val;
                j=j->next;
            }
            else    b=0;
            sum = a+b+add;
            if (sum>=10) {
                sum%=10;
                add=1;
            }
            else    add=0;
            list->next = new ListNode(sum);
            if (!answer)    answer=list->next; 
            list = list->next;
        }
        if (add)
            list->next = new ListNode(1);
        return answer;
    }
};

File: 2. Add Two Numbers/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0), node = new ListNode(0), last = node;
        head.next = node;
        while (l1 != null || l2 != null) {
            node.val += (l1 == null ? 0 : l1.val) + (l2 == null ? 0 : l2.val);
            node.next = new ListNode(node.val/10);
            node.val %= 10;
            last = node;
            node = node.next;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        if (node.val == 0)  last.next = null;
        return head.next;
    }
}
